### PR TITLE
oop-programs_with_interesting_classes-easy-q4

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# oop-programs_with_interesting_classes-easy-q4
+
+- Parsons Question
+
+## Question Text
+
+Life is a beautiful thing, and as time passes, things grow. Write a Java class about plants called `Tree`.
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:7]:
```suggestion
Life is a beautiful thing, and as time passes, things grow. Write a Java class about plants called `Tree` to model a plant with the requirements below:
```

File: src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    - A `double` attribute called **size**
+
+- Constructors:
+    - A constructor visible from another class that accepts a `String` and a `double` as parameters, and uses the values
+      to initialize the **type** and **size** attributes
+    - Note that this constructor should be visible from all other classes.
+
+- Methods:
+    - A accessor and mutator methods for the **type** attribute called `getType` and `setType`.
+    - A accessor and mutator methods for the **size** attribute called `getSize` and `setSize`.
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:25]: I think we should change this wording so we are consistent with the rest of the questions.

Comment (Hedgemon4) [reply src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:25]: Also please include the return type for the `set` methods

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static global.tools.CustomAssertions._assertEquals;
+
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class MainTest {
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:17]: Can you please add the label that this is a parsons question to the start of the test as well like we did in other categories please?

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        Object[][] constructorArgs = {
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(constructorArgs);
+        String incorrectNumDays = "The %s method in the %s class does not correctly change the %s attribute.";
+        testClass.callMethod(methodName, classInstance);
+        _assertEquals(value2 + 4, testClass.getFieldValue(classInstance, attributeName2),
+                String.format(incorrectNumDays, methodName, className, attributeName2));
+    }
```
Comment (Hedgemon4) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:105]: As this is a parsons question, I think we can just combine these two tests, so you can check the method call signature and then check that the method works.

Comment (Hedgemon4) [reply src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:105]: And the same for the other methods.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# oop-programs_with_interesting_classes-easy-q4
+
+- Parsons Question
+
+## Question Text
+
+Life is a beautiful thing, and as time passes, things grow. Write a Java class about plants called `Tree` to model a plant with the requirements below:
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:7]:
```suggestion
Life is everywhere on the planet, and it is constantly changing. Write a Java class about plants called `Tree` to model a plant with the requirements below:
```

File: src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    - Note that none of the attributes in this class should be visible from another class
+    - A `String` attribute called **type**
+    - A `double` attribute called **size**
+
+- Constructors:
+    - A constructor visible from another class that accepts a `String` and a `double` as parameters, and uses the values
+      to initialize the **type** and **size** attributes
+    - Note that this constructor should be visible from all other classes.
+
+- Methods:
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:23]: make a note of their visibility

File: src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    - A `gettingOlder` method which addes 4 to the **size** attribute.
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/QuestionText.md:24]:
```suggestion
    - A `gettingOlder` method which adds 4 to the **size** attribute.
```

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        assertTrue(testClass.hasMethod("toString", null), String.format(incorrectMethodDefinition, "toString", className));
+        assertTrue(testClass.hasModifier("toString", null, "public"), String.format(incorrectModifierMessage, "toString", className));
+        assertTrue(testClass.hasReturnType("toString", null, String.class), String.format(incorrectReturnType, "toString", className));
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] methodModifiers = {"public"};
+        String expected = "This is a " + value1 + " tree with " + value2 + " centimeters in diameter.";
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:211]: make the string a variable at the top of the test

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        assertTrue(testClass.hasModifier(methodName, null, "public"), String.format(incorrectModifierMessage, methodName, className));
+        assertTrue(testClass.hasReturnType(methodName, null, Void.TYPE), String.format(incorrectReturnType, methodName, className));
+        Object[][] constructorArgs = {
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(constructorArgs);
+        String incorrectNumDays = "The %s method in the %s class does not correctly change the %s attribute.";
+        testClass.callMethod(methodName, classInstance);
+        _assertEquals(value2 + 4, testClass.getFieldValue(classInstance, attributeName2),
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:100]: make the 4 number a variable at the top of the class

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        assertTrue(testClass.hasReturnType(getAttributeName1, null, String.class), String.format(incorrectReturnType, getAttributeName1, className));
+
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getAttributeName1, "does not correctly get the value of the", attributeName1, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:122]:
```suggestion
                "Your", getAttributeName1, "method does not correctly get the value of the", attributeName1, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+        assertTrue(testClass.hasModifier(getAttributeName2, null, "public"), String.format(incorrectModifierMessage, getAttributeName2, className));
+        assertTrue(testClass.hasReturnType(getAttributeName2, null, double.class), String.format(incorrectReturnType, getAttributeName2, className));
+        Object[][] arguments = {
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        String[] getMethodModifiers = {"public"};
+        String incorrectGetMethodMessage = String.join(" ",
+                "Your", getAttributeName2, "does not correctly get the value of the", attributeName2, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:143]:
```suggestion
                "Your", getAttributeName2, "method does not correctly get the value of the", attributeName2, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {updatedValue1, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setAttributeName1, "does not correctly set the value of the", attributeName1, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:167]:
```suggestion
                "Your", setAttributeName1, "method does not correctly set the value of the", attributeName1, "attribute.");
```

File: src/oop/programs_with_interesting_classes/easy/q4/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+                {value1, String.class},
+                {value2, double.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {updatedValue2, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setAttributeName2, "does not correctly set the value of the", attributeName2, "attribute.");
```
Comment (SamamthaH) [inline src/oop/programs_with_interesting_classes/easy/q4/MainTest.java:191]:
```suggestion
                "Your", setAttributeName2, "method does not correctly set the value of the", attributeName2, "attribute.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See Sam's review

**Pull request merged**